File: ruby/leap/leap.rb
module Year
  LEAP_YEAR = 4
  private_constant :LEAP_YEAR
  CENTURIAL_YEAR = 100
  private_constant :CENTURIAL_YEAR
  LEAP_CYCLE_LENGTH = 400
  private_constant :LEAP_CYCLE_LENGTH

  module_function

  def leap?(year)
    leap_year?(year) && (non_centurial_year?(year) || leap_cycle_year?(year))
  end

  def leap_year?(year)
    (year % LEAP_YEAR).zero?
  end
  private_class_method :leap_year?

  def non_centurial_year?(year)
    (year % CENTURIAL_YEAR).nonzero?
  end
  private_class_method :non_centurial_year?

  def leap_cycle_year?(year)
    (year % LEAP_CYCLE_LENGTH).zero?
  end
  private_class_method :leap_cycle_year?
end

File: elixir/leap/leap.exs
defmodule Year do
  @leap_year 4
  @centurial_year 100
  @leap_cycle_length @leap_year * @centurial_year

  defguardp leap?(year)
            when rem(year, @leap_year) == 0 and
                   (rem(year, @centurial_year) != 0 or
                      rem(year, @leap_cycle_length) == 0)

  @doc """
  Returns whether 'year' is a leap year.

  A leap year occurs:

  on every year that is evenly divisible by 4
    except every year that is evenly divisible by 100
      unless the year is also evenly divisible by 400
  """
  @spec leap_year?(non_neg_integer) :: boolean
  def leap_year?(year) when leap?(year), do: true
  def leap_year?(_year), do: false
end
